### PR TITLE
add `[%exclave]` support

### PR DESCRIPTION
**Pull request by riaqn**
This patch adds the support for the `[%exclave]` extension, which can be applied to an expression and results in early ending of the current region before the expression is evaluated. As such, it can only appear in the tail position of the region. A motivating example is as follows:
```
let foo (local_ x) = 
let local_ y = (complex computation on x) in 
if y then [%exclave] (local_ None)
else  [%exclave] (local_ (Some x))
```
where `foo` is supposed to be some helper function that takes arguments from parent region and returns values also in parent region. The intermediate results (`y`) it uses are stored in its local stack and released upon entering `[%exclave]`. As a result, you are not allowed to refer to `y` in `[%exclave]`.

### CODE REVIEW

File: testsuite/tests/typing-local/unregion.ml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+(* TEST
+ * expect *)
+
+(* TYPING TESTS *)
+(* TYPING TESTS *)
+(* TYPING TESTS *)
+(* the following two tests show that y is regional *)
```
Comment (lpw25) [inline testsuite/tests/typing-local/unregion.ml:7]: This comment is a little confusing. It isn't really that `y` is regional, it's that it is local for the caller's region rather than this function's region. I'd probably reword this to something like "the following two tests show that y can leave the current function despite being local".

File: testsuite/tests/typing-local/unregion.ml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+5 |     ref y
+            ^
+Error: This value escapes its region
+|}]
+
+let foo x =
+  [%unregion] (
+    let local_ y = Some x in
+    (* y is not local either *)
+    (* because it can escape *)
```
Comment (lpw25) [inline testsuite/tests/typing-local/unregion.ml:25]: This comment is a bit confusing. `y` is local it's just part of the caller's region. It would probably be better to say that `y` can escape the function even though it is local.

File: testsuite/tests/typing-local/unregion.ml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+[%%expect{|
+Line 3, characters 10-30:
+3 |   let z = [%unregion] (Some y) in
+              ^^^^^^^^^^^^^^^^^^^^
+Error: Unregion expression should only be in tail position
+|}]
+
+(* SEMANTICS TESTS *)
+(* SEMANTICS TESTS *)
+(* SEMANTICS TESTS *)
```
Comment (lpw25) [inline testsuite/tests/typing-local/unregion.ml:58]: Duplicated comment.

**Review by lpw25 — COMMENTED** [review]
A few changes needed but mostly looks good.

File: lambda/simplif.ml [hunk 4]
```diff
@@ -911,7 +925,7 @@ let simplify_local_functions lam =
         let curr_scope =
           match ap_region_close with
           | Rc_normal | Rc_nontail -> !current_scope
-          | Rc_close_at_apply -> !current_region_scope
+          | Rc_close_at_apply -> Option.get !current_region_scope  (* will raise if None *)
```
[the remark this answers is not in the document]
Comment (riaqn) [reply lambda/simplif.ml:928]: Yes should we have a dedicated exception for this?

**Comment by riaqn** [comment]
This PR has been rebased on top of version 4.14.

**Comment by riaqn** [comment]
The last commit made it so that any function ending with `[%unregion]` will always return `local_`. This is to accommodate some code in the compiler which relies on that bit to decide if the function might allocate in parent region. This also gives the expected output for a test case that wasn't working before (and I didn't know why).

**Comment by antalsz** [comment]
We should strongly consider switching to using `[@unregion]` (separately, I want to make that change for `[%local]`); in terms of making sure code is compatible with standard OCaml, `unregion` is perfectly ignorable.  Also, if we want to support an `unregion_` keyword, then we also need to update our syntax-rewriter to support `[@unregion]`; that's a job for me (as long as the location information is propagated correctly), just let me know when it's necessary.

**Comment by goldfirere** [comment]
Yes -- let's use attributes for anything that can safely be just ignored in the upstream compiler.

**Comment by riaqn** [comment]
Hmm do you think I should make the change in this PR, or do it some time later?

**Comment by goldfirere** [comment]
Antal may see things differently, but I think it's best to merge patches when they're not going to cause trouble for others. I *think* the infrastructure to handle e.g. `[@unregion]` correctly is still some time away, and so merging what you have and then quickly (this week?) changing it sounds reasonable to me.

**Comment by antalsz** [comment]
Because we can handle `[@unregion]` correctly via the refactor-based rewriter, I'd prefer making that change in ocaml-jst before we roll it out to users.  Whether that means in this PR or a different PR is something I'm happy to leave in your hands, @riaqn.  Also, are we planning to add syntax for it?

File: typing/typecore.ml [hunk 5]
```diff
@@ -374,7 +393,12 @@ let mode_argument ~funct ~index ~position ~partial_app alloc_mode =
                 Id_prim _), 1, Tail ->
      (* The second argument to (&&) and (||) is in
         tail position if the call is *)
-     mode_return (Value_mode.local_to_regional vmode)
+      (* CR zqian: this is suspicious, 
+         because vmode is not the mode of the region;
+         ideally the `position` should contain the mode 
+         of the region so we can pass it to mode_return;
+         but this might be fine because of mode-crossing? *)
+     mode_return vmode
```
Comment (riaqn) [inline typing/typecore.ml:401]: This one needs some inspection.

**Comment by riaqn** [comment]
This PR relies on https://github.com/ocaml-flambda/flambda-backend/pull/1131 to be merged safely.

File: typing/typecore.ml [hunk 6]
```diff
@@ -380,7 +407,12 @@ let mode_argument ~funct ~index ~position ~partial_app alloc_mode =
                 Id_prim _), 1, Tail ->
      (* The second argument to (&&) and (||) is in
         tail position if the call is *)
-     mode_return (Value_mode.local_to_regional vmode)
+      (* CR zqian: this is suspicious,
+         because vmode is not the mode of the region;
+         ideally the `position` should contain the mode
+         of the region so we can pass it to mode_return;
+         but this might be fine because of mode-crossing? *)
```
Comment (lpw25) [inline typing/typecore.ml:414]: I agree this is dubious. I wonder if we can treat `%sequor` and `%sequand` differently elsewhere rather than here. Maybe handling them where we handle `%ignore` in `type_application` would be easier.

File: typing/typecore.ml [hunk 7]
```diff
@@ -4143,6 +4177,38 @@ and type_expect_
...
+            With the introduction of explicit region we will need something more clever *)
+          let mode' = mode_default mode in
+          (* enforce to be local; part of the compiler depends on this bit to know
+          if it allocate in parent region. *)
+          submode ~loc ~env ~reason:Other Value_mode.local mode';
+          let env' = Env.add_exclave_lock env in
+          let exp =
+            type_expect ?in_function ~recarg env' mode' sbody ty_expected_explained
+          in
+          (* this whole thing returns regional value *)
```
Comment (lpw25) [inline typing/typecore.ml:4202]: This comment doesn't add anything

File: typing/typecore.ml [hunk 8]
```diff
@@ -4143,6 +4177,38 @@ and type_expect_
...
+          raise (Error (loc, env, Exclave_in_nontail_position))
+        | RTail (mode, _) ->
+          (* mode' is non-tail-of-region, because currently our language cannot construct
+           region in the tail of another region.
+            With the introduction of explicit region we will need something more clever *)
+          let mode' = mode_default mode in
+          (* enforce to be local; part of the compiler depends on this bit to know
+          if it allocate in parent region. *)
+          submode ~loc ~env ~reason:Other Value_mode.local mode';
+          let env' = Env.add_exclave_lock env in
```
Comment (lpw25) [inline typing/typecore.ml:4198]: `new_env` is the usual name for this

File: typing/typecore.ml [hunk 9]
```diff
@@ -4143,6 +4177,38 @@ and type_expect_
...
+          let exp =
+            type_expect ?in_function ~recarg env' mode' sbody ty_expected_explained
+          in
+          (* this whole thing returns regional value *)
+          submode ~loc ~env ~reason:Other Value_mode.regional expected_mode;
+          { exp_desc = Texp_exclave exp;
+            exp_loc = loc;
+            exp_extra = [];
+            exp_type = exp.exp_type;
+            exp_env = env;  (* use the old env *)
```
Comment (lpw25) [inline typing/typecore.ml:4208]: Not worth a comment

File: typing/typecore.ml [hunk 10]
```diff
@@ -4143,6 +4177,38 @@ and type_expect_
...
+      if (txt = "extension.exclave") && not (Language_extension.is_enabled Local) then
+          raise (Typetexp.Error (loc, Env.empty, Unsupported_extension Local));
+      begin
+        match expected_mode.position with
+        | RNontail ->
+          raise (Error (loc, env, Exclave_in_nontail_position))
+        | RTail (mode, _) ->
+          (* mode' is non-tail-of-region, because currently our language cannot construct
+           region in the tail of another region.
+            With the introduction of explicit region we will need something more clever *)
```
Comment (lpw25) [inline typing/typecore.ml:4193]: I would probably just say `RNontail` rather than "non-tail-of-region". The second line has weird indentation. s/construct region/construct a region/. The second sentence doesn't really add anything.

File: typing/typecore.ml [hunk 11]
```diff
@@ -4143,6 +4177,38 @@ and type_expect_
...
+          let mode' = mode_default mode in
+          (* enforce to be local; part of the compiler depends on this bit to know
+          if it allocate in parent region. *)
```
Comment (lpw25) [inline typing/typecore.ml:4196]: The wording here isn't very good. Perhaps:
> The middle-end relies on all functions which allocate into their parent's region having a return mode of local.

File: typing/typecore.ml [hunk 12]
```diff
@@ -4613,20 +4679,21 @@ and type_expect_
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
   | Pexp_while(scond, sbody) ->
-      let cond_env,wh_cond_region =
-        if is_local_returning_expr scond then env, false
-        else Env.add_region_lock env, true
+      let cond_env,wh_cond_region,mode=
+      (* not checking is_local_returning_expr because you should just use exclave *)
```
Comment (lpw25) [inline typing/typecore.ml:4683]: The comment only makes sense in relation to the diff. Best to remove it.

File: typing/typecore.ml [hunk 13]
```diff
@@ -4613,20 +4679,21 @@ and type_expect_
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
   | Pexp_while(scond, sbody) ->
-      let cond_env,wh_cond_region =
-        if is_local_returning_expr scond then env, false
-        else Env.add_region_lock env, true
+      let cond_env,wh_cond_region,mode=
```
Comment (lpw25) [inline typing/typecore.ml:4682]: With the `if` gone, this code can now be tidied up. For instance `wh_cond_region` is no longer needed since it is always just `true`.

File: typing/typecore.ml [hunk 14]
```diff
@@ -4613,20 +4679,21 @@ and type_expect_
...
       in
       let wh_cond =
-        type_expect cond_env mode_local scond
+        type_expect cond_env mode scond
           (mk_expected ~explanation:While_loop_conditional Predef.type_bool)
       in
-      let body_env,wh_body_region =
-        if is_local_returning_expr sbody then env, false
-        else Env.add_region_lock env, true
+      (* local because unit cross modes  *)
```
Comment (lpw25) [inline typing/typecore.ml:4690]: I think this comment is confusing. I would say that it's local because the value isn't going to escape the region. Probably just remove the comment.

File: typing/typecore.ml [hunk 15]
```diff
@@ -4613,20 +4679,21 @@ and type_expect_
...
+      let body_env,wh_body_region,position =
```
Comment (lpw25) [inline typing/typecore.ml:4691]: This can be refactored now

File: typing/typecore.ml [hunk 16]
```diff
@@ -4637,22 +4704,22 @@ and type_expect_
...
       in
       let for_id, new_env =
         type_for_loop_index ~loc ~env ~param
       in
-      let new_env, for_region =
-        if is_local_returning_expr sbody then new_env, false
-        else Env.add_region_lock new_env, true
+      let new_env, for_region, position =
+      (* local because unit cross modes  *)
+        Env.add_region_lock new_env, true, RTail (Value_mode.local, Nontail)
```
Comment (lpw25) [inline typing/typecore.ml:4719]: Same as above

**Review by lpw25 — COMMENTED** [review]
Looks pretty good. Should be good to go once various superficial things are fixed.

File: lambda/simplif.ml [hunk 17]
```diff
@@ -352,7 +355,8 @@ let simplify_exits lam =
   | Lifused(v, l) -> Lifused (v,simplif ~layout ~try_depth l)
   | Lregion (l, ly) -> Lregion (
       simplif ~layout ~try_depth:(try_depth + 1) l,
-      result_layout ly)
+      result_layout ly) (* lwhite suggest to double check this case *)
```
Comment (lpw25) [inline lambda/simplif.ml:358]: Comment no longer needed

File: testsuite/tests/typing-local/loop_regions.ml [hunk 18]
```diff
@@ -98,12 +101,43 @@ let nonloc_while_cond () =
...
+let loc_func () =
+  let offset_func = ref (-1) in
+  let fun_exclave r =
+    [%exclave] (
+        let z = local_ (Some (Sys.opaque_identity 42)) in
+        let _ = (opaque_local z) in
+        r := local_stack_offset ()
+    ) in
+  let offset1 = local_stack_offset () in
+
```
Comment (lpw25) [inline testsuite/tests/typing-local/loop_regions.ml:114]: Weird blank line

File: testsuite/tests/typing-local/loop_regions.ml [hunk 19]
```diff
@@ -98,12 +101,43 @@ let nonloc_while_cond () =
...
+let nonloc_func () =
+  let offset_func = ref (-1) in
+  let fun_nonexclave r =
+    let z = local_ (Some (Sys.opaque_identity 42)) in
+    let _ = (opaque_local z) in
+    r := local_stack_offset ()
+  in
+  let offset1 = local_stack_offset () in
+
```
Comment (lpw25) [inline testsuite/tests/typing-local/loop_regions.ml:128]: Weird blank line

File: typing/typecore.ml [hunk 20]
```diff
@@ -4679,25 +4673,23 @@ and type_expect_
         exp_attributes = sexp.pexp_attributes;
         exp_env = env }
   | Pexp_while(scond, sbody) ->
-      let cond_env,wh_cond_region,mode=
-      (* not checking is_local_returning_expr because you should just use exclave *)
-        Env.add_region_lock env, true, mode_region (Value_mode.local)
+      let cond_env,mode=
+        Env.add_region_lock env, mode_region Value_mode.local
```
Comment (lpw25) [inline typing/typecore.ml:4677]: No need for this to be a single `let` with a tuple, just have two `let`s

File: typing/typecore.ml [hunk 21]
```diff
@@ -4679,25 +4673,23 @@ and type_expect_
...
       in
       let wh_cond =
         type_expect cond_env mode scond
           (mk_expected ~explanation:While_loop_conditional Predef.type_bool)
       in
-      (* local because unit cross modes  *)
-      let body_env,wh_body_region,position =
-        Env.add_region_lock env, true, RTail (Value_mode.local, Nontail)
+      let body_env,position =
+        Env.add_region_lock env, RTail (Value_mode.local, Nontail)
```
Comment (lpw25) [inline typing/typecore.ml:4684]: Same as above

File: typing/typecore.ml [hunk 22]
```diff
@@ -4714,16 +4706,15 @@ and type_expect_
       let for_id, new_env =
         type_for_loop_index ~loc ~env ~param
       in
-      let new_env, for_region, position =
-      (* local because unit cross modes  *)
-        Env.add_region_lock new_env, true, RTail (Value_mode.local, Nontail)
+      let new_env, position =
+        Env.add_region_lock new_env, RTail (Value_mode.local, Nontail)
```
Comment (lpw25) [inline typing/typecore.ml:4710]: And this one

File: jane/doc/local-reference.md [hunk 23]
```diff
@@ -442,39 +442,81 @@ be the caller's (or that of the caller's caller, etc., if the caller is also
...
+  | [] -> Ok { res = 0 }
+  | x :: xs ->
+      match p x with
+      | false -> None
+      | true ->
+          match count_result f xs with
+          | None as r -> r
+          | Some { res = count } ->
+              Some { res = count + 1 }
+```
```
Comment (lpw25) [inline jane/doc/local-reference.md:497]: Let's just take this example out for now. I'll add a better one in a later PR.

File: jane/doc/local-reference.md [hunk 24]
```diff
@@ -442,39 +442,81 @@ be the caller's (or that of the caller's caller, etc., if the caller is also
...
+in which case the result of the entire function would be `None`. It is designed
+not to allocate heap memory, instead using the stack for all `Some` allocations.
+However, it will currently use O(N) stack space because all allocations occur in
+the original caller's stack frame. To improve its space usage, we remove the
+`local_` annotation (so the function has its own region), and wrap `Some {res =
+count + 1}` inside `[%exclave]` to release the region before the allocation.
+After the revision, the function should use O(1) stack space.
+
+
+`[%exclave]` terminates the current region, so values within it cannot be used
```
Comment (lpw25) [inline jane/doc/local-reference.md:515]:
```suggestion
`[%exclave]` terminates the current region, so local values from that region cannot be used
```

**Review by lpw25 — APPROVED** [review]
Small small suggestions then this is good to go

File: jane/doc/local-reference.md (hunk 23 above)
Comment (riaqn) [reply jane/doc/local-reference.md:497]: is it better to have a silly but working example, than to have no example?
